src/encoding/src/rlp.cairo

```diff
@@ -0,0 +1,197 @@
+use alexandria_data_structures::array_ext::ArrayTraitExt;
+use alexandria_numeric::integers::UIntBytes;
+
+// Possible RLP errors
+#[derive(Drop, Copy, PartialEq)]
+enum RLPError {
+    EmptyInput,
+    InputTooShort
+}
+
+// Possible RLP types
+#[derive(Drop, PartialEq)]
+enum RLPType {
+    String,
+    List
+}
+
+#[derive(Drop, Copy, PartialEq)]
+enum RLPItem {
+    String: Span<u8>,
+    List: Span<RLPItem>
+}
+
+#[generate_trait]
+impl RLPImpl of RLPTrait {
+    /// Returns RLPType from the leading byte with
+    /// its offset in the array as well as its size.
+    ///
+    /// # Arguments
+    /// * `input` - Array of byte to decode
+    /// # Returns
+    /// * `(RLPType, offset, size)` - A tuple containing the RLPType
+    /// the offset and the size of the RLPItem to decode
+    /// # Errors
+    /// * empty input - if the input is empty
+    /// * input too short - if the input is too short for a given
+    fn decode_type(input: Span<u8>) -> Result<(RLPType, u32, u32), RLPError> {
+        let input_len = input.len();
+        if input_len == 0 {
+            return Result::Err(RLPError::EmptyInput(()));
+        }
+
+        let prefix_byte = *input[0];
+
+        if prefix_byte < 0x80 { // Char
+            return Result::Ok((RLPType::String, 0, 1));
+        } else if prefix_byte < 0xb8 { // Short String
+            return Result::Ok((RLPType::String, 1, prefix_byte.into() - 0x80));
+        } else if prefix_byte < 0xc0 { // Long String
+            let len_bytes_count: u32 = (prefix_byte - 0xb7).into();
+            if input_len <= len_bytes_count {
+                return Result::Err(RLPError::InputTooShort(()));
+            }
+            let string_len_bytes = input.slice(1, len_bytes_count);
+            let string_len: u32 = UIntBytes::from_bytes(string_len_bytes).unwrap();
```

Comment:
My very bad, I missed those !!

Reply:
Having to justify the safety of those unwrap made me realize they weren't actually safe.
The problem being that the payload length can be on 8 Bytes ( therefore an u64 ). But arrays don't seem to be able to hold more than MAXU32 values. ( I am apparently unable to test this with my computers :s )

I therefore added a check on the amount of bytes representing the payload length and currently return an error if the payload is > 4 bytes.
But I am not convinced by this because it doesn't allow us to decode very long payloads. I ignore how frequently we are supposed to decode payload with a length above MAXU32 in real life but still, it isn't compliant with the official RLP doc and I don't know how to make it compliant right now using vanilla arrays/spans.

I hope I'm being clear, if I sound confusing, feel free to reach me on TG/Disc/whatever, I'd be very glad to explain the problem !

Reply:
you can use a 2d array

Reply:
but maybe this should be in another pr idk wdyt ?